### PR TITLE
Implement `Arbitrary` for the set of `GearCall`s

### LINKED ISSUE #2314: Implement `Arbitrary` on `GearCall`
### Proposal

To make fuzzer smarted it should fully control the executed input to shrink/mutate it properly, as shrinking u64 seed is impossible. So by implementing `Arbitrary` on `GearCall` we are giving more control over inputs to the fuzzer.

### PR DESCRIPTION
**Pull request by SabaunT**
Resolves #2314

Significant step for upcoming precise runtime-fuzzing.

- Injected `log::trace` calls all over `gear-wasm-gen` to make easier diagnostics of `wasm-gen` and fuzzing. 
- Introduce `GearCalls` wrapper over the collection of `GearCall`, that implements `Arbitrary`.
- Rename valid configs bundle from `ValidGearWasmConfigsBundle` to `StandardGearWasmConfigsBundle`.

@gear-tech/dev

### CODE REVIEW

File: utils/call-gen/src/arbitrary_call.rs [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+/// Maximum payload size for the fuzzer - 512 KiB.
+const MAX_PAYLOAD_SIZE: usize = 512 * 1024;
+static_assertions::const_assert!(MAX_PAYLOAD_SIZE <= gear_core::message::MAX_PAYLOAD_SIZE);
+
+#[derive(Debug, Clone)]
+/// New-type wrapper over array of [`GearCall`]s.
+///
+/// It's main purpose is to be an implementor of `Arbitrary` for the array of [`GearCall`]s.
+/// New-type is required as array is always a foreign type.
+pub struct GearCalls(pub [GearCall; GearCalls::MAX_CALLS]);
```
Comment (StackOverflowExcept1on) [inline utils/call-gen/src/arbitrary_call.rs:40]: reorder `///` and `derive`

File: utils/call-gen/src/arbitrary_call.rs [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+            let payload = arbitrary_payload(u)?;
+            log::trace!(
+                "Random data after payload (upload_program) gen {}, iter - {i}",
+                u.len()
+            );
+            log::trace!("Payload (upload_program) length {:?}", payload.len());
+
+            programs[i] = ProgramId::generate(CodeId::generate(&code), &salt);
+            calls[i] = MaybeUninit::new(GearCall::UploadProgram(UploadProgramArgs((
```
Comment (StackOverflowExcept1on) [inline utils/call-gen/src/arbitrary_call.rs:100]: `calls[i].write(GearCall::UploadProgram(UploadProgramArgs(...))` here and on other places

File: utils/call-gen/src/arbitrary_call.rs (hunk 1 above)
Comment (SabaunT) [reply utils/call-gen/src/arbitrary_call.rs]: done

File: utils/call-gen/src/arbitrary_call.rs (hunk 2 above)
Comment (SabaunT) [reply utils/call-gen/src/arbitrary_call.rs]: done

File: utils/wasm-gen/src/generator/syscalls/imports.rs [hunk 3]
```diff
@@ -343,6 +357,8 @@ impl<'a> SysCallsImportsGenerator<'a> {
             (module_builder.build(), idx)
         });
 
+        log::trace!("Builded proper reservation send call");
```
Comment (mertwole) [inline utils/wasm-gen/src/generator/syscalls/imports.rs:360]:
```suggestion
        log::trace!("Built proper reservation send call");
```

**Review by mertwole — CHANGES_REQUESTED** [review]

File: utils/call-gen/src/arbitrary_call.rs [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        if u.len() < 25_000_000_usize {
+            log::trace!("Not enough bytes for creating gear calls");
+            return Err(arbitrary::Error::NotEnoughData);
+        }
+
+        let log_data = format!(
+            "Generated from corpus - {}",
+            get_sha1_string(u.peek_bytes(u.len()).expect("checked"))
+        );
+        let gas = 249_000_000_000_u64;
```
[the remark this answers is not in the document]
Comment (SabaunT) [reply utils/call-gen/src/arbitrary_call.rs:69]: `GasLimit` is provided from runtime. We don't have an access to runtime constant from `call-gen`. Seems no need to.

File: utils/call-gen/src/arbitrary_call.rs [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+
+        log::trace!("New GearCalls generation: random data received {}", u.len());
+
+        // 25 MiB is an approximate assessment for 35 calls,
+        // where each call of:
+        // 1. `UploadProgram` requires 1024 KiB for payload an salt and 50 KiB for code.
+        // 2. `SendMessage` requires 512 Kib for payload.
+        // So, [10 * (1024 + 50)] + [25 * 512] = 23540 KiB
+        if u.len() < 25_000_000_usize {
+            log::trace!("Not enough bytes for creating gear calls");
```
[the remark this answers is not in the document]
Comment (SabaunT) [reply utils/call-gen/src/arbitrary_call.rs:61]: That's not an error. Also this info must be logged on lower levels, like `trace`. It's needed to do a deep diagnostics of fuzzer. Usually trace level is off.

File: utils/wasm-gen/src/config/syscalls/param.rs [hunk 6]
```diff
@@ -55,7 +55,7 @@ impl Default for SysCallsParamsConfig {
                 (ParamType::Gas, (0..=250_000_000_000).into()),
                 (ParamType::MessagePosition, (0..=10).into()),
                 (ParamType::Duration, (0..=100).into()),
-                (ParamType::Delay, (0..=100).into()),
+                (ParamType::Delay, (1..=4).into()),
```
Comment (breathx) [inline utils/wasm-gen/src/config/syscalls/param.rs:58]: why start from 1?

**Review by breathx — APPROVED** [review]

File: utils/wasm-gen/src/generator/syscalls/additional_data.rs [hunk 7]
```diff
@@ -181,13 +189,25 @@ impl<'a> AdditionalDataInjector<'a> {
...
 
         let export_idx = self
             .module
             .gear_entry_point(EntryPointName::Init)
-            .or_else(|| self.module.gear_entry_point(EntryPointName::Handle))
-            .or_else(|| self.module.gear_entry_point(EntryPointName::HandleReply))
+            .map(|idx| {
+                log::trace!("Info will be logged in init");
+                idx
+            })
```
[the remark this answers is not in the document]
Comment (SabaunT) [reply utils/wasm-gen/src/generator/syscalls/additional_data.rs:202]: this is experimental api

File: utils/wasm-gen/src/config/syscalls/param.rs (hunk 6 above)
Comment (SabaunT) [reply utils/wasm-gen/src/config/syscalls/param.rs:58]: Zero duration is not allowed, isn't it?

File: utils/call-gen/src/arbitrary_call.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (SabaunT) [reply utils/call-gen/src/arbitrary_call.rs:69]: or you mean to introduce a constant `GAS_LIMIT` in `gear_call_gen` and do `GAS_LIMIT * 99 / 100`?

File: utils/call-gen/src/arbitrary_call.rs [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+        // # Safety:
+        //
+        // Called when gear calls are initialized. Transmute the array to the
+        // initialized type.
+        mem::transmute::<_, [GearCall; GearCalls::MAX_CALLS]>(uninit_calls)
+    }
+}
+
+fn config(
+    programs: [ProgramId; 10],
```
[the remark this answers is not in the document]
Comment (SabaunT) [reply utils/call-gen/src/arbitrary_call.rs]: done

File: utils/wasm-gen/src/generator/syscalls/imports.rs (hunk 3 above)
Comment (SabaunT) [reply utils/wasm-gen/src/generator/syscalls/imports.rs]: done

File: utils/call-gen/src/arbitrary_call.rs [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+        // # Safety:
+        //
+        // Create an uninitialized array of `MaybeUninit`. The `assume_init` is
+        // safe because the type we are claiming to have initialized here is a
+        // bunch of `MaybeUninit`s, which do not require initialization.
+        MaybeUninit::uninit().assume_init()
+    }
+}
+
+fn transmute_calls(
```
[the remark this answers is not in the document]
Comment (SabaunT) [reply utils/call-gen/src/arbitrary_call.rs]: done

File: utils/call-gen/src/arbitrary_call.rs [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+            log::trace!("Random data before wasm gen {}, iter - {i}", u.len());
+
+            let code = gear_wasm_gen::generate_gear_program_code(
+                u,
+                config(programs, Some(log_data.clone())),
+            )?;
+            log::trace!("Random data after wasm gen {}, iter - {i}", u.len());
+            log::trace!("Code length {:?}", code.len());
+
+            let salt = arbitrary_payload(u)?;
```
[the remark this answers is not in the document]
Comment (SabaunT) [reply utils/call-gen/src/arbitrary_call.rs]: done

**Review by StackOverflowExcept1on — APPROVED** [review]

File: utils/call-gen/src/arbitrary_call.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (mertwole) [reply utils/call-gen/src/arbitrary_call.rs:69]: you can introduce constant to not use magic numbers

**Review by mertwole — APPROVED** [review]

File: utils/wasm-gen/src/config/syscalls/param.rs (hunk 6 above)
Comment (breathx) [reply utils/wasm-gen/src/config/syscalls/param.rs:58]: > Zero duration is not allowed, isn't it?

It's delay, not duration. Duration exactly can't be started from 0

Comment (SabaunT) [reply utils/wasm-gen/src/config/syscalls/param.rs]: done

File: utils/call-gen/src/arbitrary_call.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (SabaunT) [reply utils/call-gen/src/arbitrary_call.rs]: done

**Pull request merged**